examples/github-deploy.yml

```diff
@@ -0,0 +1,80 @@
+name: Build and upload to PyPI
+
+on: [push, pull_request, release]
+
+jobs:
+  build_wheels:
+    name: Build wheels on ${{ matrix.os }}
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-18.04, windows-latest, macos-latest]
+
+    steps:
+    - uses: actions/checkout@v2
+
+    - uses: actions/setup-python@v2
+      name: Install Python
+      with:
+        python-version: '3.7'
+
+    - name: Install cibuildwheel
+      run: |
+        python -m pip install cibuildwheel==1.4.2
+
+    - name: Install Visual C++ for Python 2.7
+      if: startsWith(matrix.os, 'windows')
+      run: |
+        choco install vcpython27 -f -y
+
+    - name: Build wheels
+      run: |
+        python -m cibuildwheel --output-dir wheelhouse
+
+    - uses: actions/upload-artifact@v2
+      with:
+        path: ./wheelhouse/*.whl
+
+  build_sdist:
+    name: Build source distribution
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+
+    - uses: actions/setup-python@v2
+      name: Install Python
+      with:
+        python-version: '3.7'
+
+    - name: Build sdist
+      run: python setup.py sdist
+
+    - uses: actions/upload-artifact@v2
+      with:
+        path: dist/*.tar.gz
+
+  upload_pypi:
+    needs: [build_wheels, build_sdist]
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/setup-python@v2
+      with:
+        python-version: "3.8"
+
+    - name: Install dependencies
+      run: python -m pip install twine
+
+    - uses: actions/download-artifact@v2
+      with:
+        name: artifact
+        path: dist
+
+    - run: twine check dist/*
+
+    - uses: pypa/gh-action-pypi-publish@master
+      with:
+        user: __token__
+        password: ${{ secrets.pypi_password }}
+        # To test: repository_url: https://test.pypi.org/legacy/
+      if: github.event_name == 'release' && github.event.action == 'published'
```

Comment:
this workflow triggers on every push, pull, release : 
https://github.com/joerick/cibuildwheel/pull/362/files#diff-b4971e357cad581b8716b72ee38f58a5R3 

...which probably does not make sense, only tagged push & a release. 

Then this action filters only to "Release published"

Reply:
I think people generally want CI to check that things build and to test resulting wheels. That's not what I'm asking - i'm wondering if pushing a tag creates a release in Github - such that this `if` statement evaluates to true.

Reply:
> i'm wondering if pushing a tag creates a release in Github 

I think no, you have to have another action for that. Tag will show just as Tags.

Reply:
No, this should only release if you create a GitHub release. It will build on tags, since that's included by default in push, I think. 

```yaml
on:
  release:
  push:
    tags:
    - 'v*'
...
```

Will certainly build on both releases and tags. That way, you can tag, download and test the wheels, then release when you are ready. If you want to release on tags, this will push to PyPY even if you don't make a GitHub release.

```yaml
if: github.event_name == 'push' && startsWith(github.event.ref, 'refs/tags')
```

Unlike Azure, there is no user interface for manually building so you can check wheels, so having the release mechanism stands in for that, and reduces your chances of making a bad release to PyPI. Depends on what you want to do.

Reply:
Like azure, though, if you upload artifacts, there's an option in the web interface to download them and play with them locally.

Reply:
I've also seen

```yaml
on: create
if: github.event_name == 'create' && github.event.ref_type == 'tag'
```

Suggested for tags, never tried it though.

Reply:
Ah, cool. Gotcha. I'm more familiar with the tag-based approach, and I _think_ that is more common. Perhaps we could provide the user with a choice here? 

```suggestion
      # upload to PyPI on every tag starting with 'v'
      if: github.event_name == 'push' && startsWith(github.event.ref, 'refs/tags/v')
      # alternatively, to publish when a GitHub Release is created, use the following rule:
      # if: github.event_name == 'release' && github.event.action == 'published'
```

Reply:
That's fine - though I'd mention that `on: release` is needed too in the comment.

It's harder to do release based releases in non-GH tools, since they have to use the GitHub API to detect releases, while GHA has easy access to it.

Reply:
Sounds all good to me